src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -6006,14 +5974,39 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     }
 
     def typed(tree: Tree, mode: Mode, pt: Type): Tree = {
+      // Extract and insert stabilizing ValDefs (if any) which might have been
+      // introduced during the typing of the original expression.
+      @inline def insertStabilizer(doTyped: Mode => Tree): Tree = {
+        if (phase.erasedTypes || mode.in(APPSELmode) || isMacroImplRef(tree)) doTyped(mode)
+        else tree match {
+          case _: Select | _: Apply | _: TypeApply =>
+            val saved = context.pendingStabilizers
+            context.pendingStabilizers = Nil
+            val (res, vdefs) =
+              try {
+                val r = doTyped(mode | APPSELmode)
+                (r, context.pendingStabilizers)
+              }
```

Comment:
I hate to be the abstraction Grinch (again!), but here goes.

I'd prefer to avoid the Tuple2 here.

More generally, the `insertStabilzer` combinator doesn't seem to be inlining as well as we'd like. I'm seeing `Mode` boxing and `Function.apply` calls in:

```
   L10
    LINENUMBER 5983 L10
    ALOAD 0
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Typers$Typer.context ()Lscala/tools/nsc/typechecker/Contexts$Context;
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Contexts$Context.pendingStabilizers ()Lscala/collection/immutable/List;
   L11
    ASTORE 7
   L12
    LINENUMBER 5984 L12
    ALOAD 0
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Typers$Typer.context ()Lscala/tools/nsc/typechecker/Contexts$Context;
    GETSTATIC scala/collection/immutable/Nil$.MODULE$ : Lscala/collection/immutable/Nil$;
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Contexts$Context.pendingStabilizers_$eq (Lscala/collection/immutable/List;)V
   L0
    LINENUMBER 5987 L0
    ALOAD 1
    NEW scala/reflect/internal/Mode
    DUP
    GETSTATIC scala/reflect/internal/Mode$.MODULE$ : Lscala/reflect/internal/Mode$;
    POP
    GETSTATIC scala/tools/nsc/package$.MODULE$ : Lscala/tools/nsc/package$;
    INVOKEVIRTUAL scala/tools/nsc/package$.Mode ()Lscala/reflect/internal/Mode$;
    INVOKEVIRTUAL scala/reflect/internal/Mode$.APPSELmode ()I
    ISTORE 12
   L13
    ILOAD 2
    ILOAD 12
    IOR
   L14
    INVOKESPECIAL scala/reflect/internal/Mode.<init> (I)V
    INVOKEINTERFACE scala/Function1.apply (Ljava/lang/Object;)Ljava/lang/Object; (itf)
    CHECKCAST scala/reflect/internal/Trees$Tree
   L15
    ASTORE 8
   L16
    LINENUMBER 5988 L16
    ALOAD 0
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Typers$Typer.context ()Lscala/tools/nsc/typechecker/Contexts$Context;
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Contexts$Context.pendingStabilizers ()Lscala/collection/immutable/List;
    ASTORE 14
   L17
    GOTO L18
   L1
   FRAME FULL [scala/tools/nsc/typechecker/Typers$Typer scala/Function1 I scala/reflect/internal/Trees$Tree scala/reflect/internal/Types$Type T I scala/collection/immutable/List] [java/lang/Throwable]
    ASTORE 9
   L19
    LINENUMBER 5991 L19
    ALOAD 0
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Typers$Typer.context ()Lscala/tools/nsc/typechecker/Contexts$Context;
    ALOAD 7
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Contexts$Context.pendingStabilizers_$eq (Lscala/collection/immutable/List;)V
    ALOAD 9
    ATHROW
   L18
   FRAME FULL [scala/tools/nsc/typechecker/Typers$Typer scala/Function1 I scala/reflect/internal/Trees$Tree scala/reflect/internal/Types$Type T I scala/collection/immutable/List scala/reflect/internal/Trees$Tree T T T
I T scala/collection/immutable/List] []
    ALOAD 0
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Typers$Typer.context ()Lscala/tools/nsc/typechecker/Contexts$Context;
    ALOAD 7
    INVOKEVIRTUAL scala/tools/nsc/typechecker/Contexts$Context.pendingStabilizers_$eq (Lscala/collection/immutable/List;)V
```

Might I suggest something like: https://github.com/retronym/scala/pull/92

Reply:
Could be the exception handler (try-finally) that breaks inlining / closure elimination